### PR TITLE
haskell: Inline static overlay

### PR DESCRIPTION
**Pull request by Ericson2314**
###### Motivation for this change

We don't force `enableStaticLibraries` for Windows or WASM because it
will just fail an assertion.

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/development/haskell-modules/generic-builder.nix [hunk 1]
```diff
@@ -33,9 +33,9 @@ in
 , profilingDetail ? "exported-functions"
 # TODO enable shared libs for cross-compiling
 , enableSharedExecutables ? false
-, enableSharedLibraries ? (ghc.enableShared or false)
+, enableSharedLibraries ? !stdenv.hostPlatform.isStatic && (ghc.enableShared or false)
 , enableDeadCodeElimination ? (!stdenv.isDarwin)  # TODO: use -dead_strip for darwin
-, enableStaticLibraries ? !(stdenv.hostPlatform.isWindows or stdenv.hostPlatform.isWasm)
```
Comment (cdepillabout) [inline pkgs/development/haskell-modules/generic-builder.nix:38]: Is this change necessary?  I think [`!(a or b) == !a and !b`](https://en.wikipedia.org/wiki/De_Morgan%27s_laws).

Although it would be really nice to have a comment above this line explaining why static libraries are disabled when the host platform is windows or wasm.

File: pkgs/development/haskell-modules/generic-builder.nix [hunk 2]
```diff
@@ -33,9 +33,9 @@ in
 , profilingDetail ? "exported-functions"
 # TODO enable shared libs for cross-compiling
 , enableSharedExecutables ? false
-, enableSharedLibraries ? (ghc.enableShared or false)
+, enableSharedLibraries ? !stdenv.hostPlatform.isStatic && (ghc.enableShared or false)
```
Comment (cdepillabout) [inline pkgs/development/haskell-modules/generic-builder.nix:36]: It might be nice to have a quick note here explaining that `hostPlatform.isStatic` means that we don't want any shared libraries built (or maybe there are not available to link against?).

(Actually I don't know this, I'm just guessing from the names... which is why it would be nice to have a comment about it.)

**Review by cdepillabout — CHANGES_REQUESTED** [review]
@Ericson2314 Could you explain what this PR does?  I don't really understand what is going on here.

Also, how would I go about testing that this PR is working as intended?

cc @nh2 as well.

File: pkgs/development/haskell-modules/generic-builder.nix (hunk 1 above)
Comment (Ericson2314) [reply pkgs/development/haskell-modules/generic-builder.nix:38]: Yes I thought i was going to change it more but then I didn't. I'll put back.

File: pkgs/development/haskell-modules/generic-builder.nix (hunk 2 above)
Comment (Ericson2314) [reply pkgs/development/haskell-modules/generic-builder.nix:36]: @cdepillabout I thought a bit about out; since this is always what `isStatic` is for I think it would be better to add more documention where it is defined instead, rather than at every use site.
